fdb-record-layer-lucene/src/main/java/org/apache/lucene/index/StandardDirectoryReaderOptimization.java

```diff
@@ -88,4 +91,60 @@ protected DirectoryReader doBody(String segmentFileName) throws IOException {
             }
         }.run(commit);
     }
+
+    static class SegmentReaderOpener {
+        private final SegmentReader[] segmentReaders;
+        private final int readerIndex;
+
+        private final SegmentCommitInfo segmentCommitInfo;
+        private final int createdVersionMajor;
+        private final IOContext ioContext;
+
+        private IOException ioException;
+        private CompletableFuture<Void> future;
+
+        SegmentReaderOpener(final SegmentReader[] segmentReaders, final int readerIndex,
+                            final SegmentCommitInfo segmentCommitInfo, final int createdVersionMajor, final IOContext ioContext) {
+            this.segmentReaders = segmentReaders;
+            this.readerIndex = readerIndex;
+            this.segmentCommitInfo = segmentCommitInfo;
+            this.createdVersionMajor = createdVersionMajor;
+            this.ioContext = ioContext;
+        }
+
+        public void start(final Executor executor) {
+            try {
+                future = CompletableFuture.supplyAsync(() -> {
+                    try {
+                        open();
+                    } catch (IOException ex) {
+                        ioException = ex;
+                    } catch (RejectedExecutionException ex) {
+                        // This happens when trying to block in openSchema and having run out of extra threads.
+                        segmentReaders[readerIndex] = null;
+                    }
+                    return null;
+                }, executor);
+            } catch (RejectedExecutionException ex) {
+                future = null;
+            }
+        }
+
+        @SuppressWarnings("PMD.PreserveStackTrace")
+        public void finish() throws IOException {
+            if (future != null) {
+                future.join();
+            }
+            if (ioException != null) {
+                throw ioException;
+            }
+            if (segmentReaders[readerIndex] == null) {
+                open();
+            }
+        }
+
+        private void open() throws IOException {
```

Comment:
Would it be easy / beneficial to add a test that runs with the `LUCENE_OPEN_PARALLELISM` and/or `executor`'s parallelism set to very low values (like 1 or 2)?
The issues are most likely to happen when you have hundreds or thousands of segments, but perhaps we can simulate by lowering the parallelism a bunch.

Reply:
Yeah. I added a test that does it with parallelism `2` after (effectively) disabling auto-merge. Not much of a stress test, but just in case all other tests happened to be below `16`.